### PR TITLE
Unify Error Handling

### PR DESCRIPTION
**Pull request by Wind4Greg**
This PR addresses issue https://github.com/w3c/vc-di-ecdsa/issues/63. It unifies error codes, error handling language, and adds error codes where needed.

***
<a href="https://pr-preview.s3.amazonaws.com/Wind4Greg/vc-di-ecdsa/pull/64.html" title="Last updated on Jun 28, 2024, 7:09 PM UTC (671150e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-di-ecdsa/64/b0ec946...Wind4Greg:671150e.html" title="Last updated on Jun 28, 2024, 7:09 PM UTC (671150e)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1148,13 +1150,14 @@ <h4>Proof Configuration (ecdsa-jcs-2019)</h4>
...
-`INVALID_PROOF_CONFIGURATION` error MUST be raised.
+|proofConfig|.|cryptosuite| is not set to `ecdsa-jcs-2019`,
+an error MUST be raised and SHOULD convey an error type of
+`PROOF_GENERATION_ERROR`.
             </li>
             <li>
 If |proofConfig|.|created| is set and if the value is not a
-valid [[XMLSCHEMA11-2]] datetime, an `INVALID_PROOF_DATETIME` error MUST be
-raised.
+valid [[XMLSCHEMA11-2]] datetime, an an error MUST be raised and SHOULD convey
```
Comment (dlongley) [inline index.html:1159]:
```suggestion
valid [[XMLSCHEMA11-2]] datetime, an error MUST be raised and SHOULD convey
```

File: index.html [hunk 2]
```diff
@@ -732,8 +732,9 @@ <h4>Transformation (ecdsa-rdfc-2019)</h4>
             <li>
 If |options|.|type| is not set to the string
 `DataIntegrityProof` and |options|.|cryptosuite| is not
-set to the string `ecdsa-rdfc-2019` then a `PROOF_TRANSFORMATION_ERROR` MUST be
-raised.
+set to the string `ecdsa-rdfc-2019` then
```
Comment (TallTed) [inline index.html:735]:
```suggestion
set to the string `ecdsa-rdfc-2019`,
```

File: index.html [hunk 3]
```diff
@@ -1060,8 +1062,8 @@ <h4>Transformation (ecdsa-jcs-2019)</h4>
             <li>
 If |options|.|type| is not set to the string
 `DataIntegrityProof` and |options|.|cryptosuite| is not
-set to the string `ecdsa-jcs-2019`, then a `PROOF_TRANSFORMATION_ERROR` MUST be
-raised.
+set to the string `ecdsa-jcs-2019`, then an error MUST be raised and SHOULD
```
Comment (TallTed) [inline index.html:1065]:
```suggestion
set to the string `ecdsa-jcs-2019`, an error MUST be raised and SHOULD
```

File: index.html (hunk 1 above)
Comment (TallTed) [inline index.html:1159]:
```suggestion
valid [[XMLSCHEMA11-2]] datetime, an error MUST be raised and SHOULD convey
```

**Review by dlongley — APPROVED** [review]

File: index.html [hunk 4]
```diff
@@ -2198,15 +2203,17 @@ <h4>parseBaseProofValue</h4>
           <ol class="algorithm">
             <li>
 Ensure the |proofValue| string starts with `u`, indicating that it is a
```
Comment (TallTed) [inline index.html:2205]:
```suggestion
```

File: index.html [hunk 5]
```diff
@@ -2198,15 +2203,17 @@ <h4>parseBaseProofValue</h4>
...
-multibase-base64url-no-pad-encoded value, throwing an error if it does not.
+multibase-base64url-no-pad-encoded value, if it does not an error MUST be raised
+and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```
Comment (TallTed) [inline index.html:2207]:
```suggestion
If the |proofValue| string does not start with `u`, indicating that it is
a multibase-base64url-no-pad-encoded value, an error MUST be raised
and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```

File: index.html [hunk 6]
```diff
@@ -2198,15 +2203,17 @@ <h4>parseBaseProofValue</h4>
...
             </li>
             <li>
 Initialize |decodedProofValue| to the result of base64url-no-pad-decoding the
 substring after the leading `u` in |proofValue|.
             </li>
             <li>
 Ensure that the |decodedProofValue| starts with the ECDSA-SD base proof header
```
Comment (TallTed) [inline index.html:2214]:
```suggestion
```

File: index.html [hunk 7]
```diff
@@ -2198,15 +2203,17 @@ <h4>parseBaseProofValue</h4>
...
-bytes 0xd9, 0x5d, and 0x00, throwing an error if it does not.
+bytes 0xd9, 0x5d, and 0x00, if it does not an error MUST be raised and SHOULD
+convey an error type of `PROOF_VERIFICATION_ERROR`.
```
Comment (TallTed) [inline index.html:2216]:
```suggestion
If the |decodedProofValue| does not start with the ECDSA-SD base proof
header bytes `0xd9`, `0x5d`, and `0x00`, an error MUST be raised and SHOULD
convey an error type of `PROOF_VERIFICATION_ERROR`.
```

File: index.html [hunk 8]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
           <ol class="algorithm">
             <li>
 Ensure the |proofValue| string starts with `u`, indicating that it is a
```
Comment (TallTed) [inline index.html:2478]:
```suggestion
```

File: index.html [hunk 9]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
...
-multibase-base64url-no-pad-encoded value, throwing an error if it does not.
+multibase-base64url-no-pad-encoded value, if it does not an error MUST be raised
+and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```
Comment (TallTed) [inline index.html:2480]:
```suggestion
If the |proofValue| string does not start with `u`, indicating that it is a
multibase-base64url-no-pad-encoded value, an error MUST be raised
and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```

File: index.html [hunk 10]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
...
             </li>
             <li>
 Initialize |decodedProofValue| to the result of base64url-no-pad-decoding the
 substring after the leading `u` in |proofValue|.
             </li>
             <li>
 Ensure that the |decodedProofValue| starts with the ECDSA-SD disclosure proof
```
Comment (TallTed) [inline index.html:2487]:
```suggestion
```

File: index.html [hunk 11]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
...
-header bytes `0xd9`, `0x5d`, and `0x01`, throwing an error if it does not.
+header bytes `0xd9`, `0x5d`, and `0x01`, if it does not an error MUST be raised
+and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```
Comment (TallTed) [inline index.html:2489]:
```suggestion
If the |decodedProofValue| does not start with the ECDSA-SD disclosure proof
header bytes `0xd9`, `0x5d`, and `0x01`, an error MUST be raised
and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```

File: index.html [hunk 12]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
...
             </li>
             <li>
 Initialize |components| to an array that is the result of CBOR-decoding the
 bytes that follow the three-byte ECDSA-SD disclosure proof header. Ensure the
 result is an array of five elements. Ensure the result is an array of five
 elements: a byte array of length 64, a byte array of length 36, an array of byte
 arrays, each of length 64, a map of integers to byte arrays of length 32, and an
```
Comment (TallTed) [inline index.html:2496]: In the `map of integers to byte arrays`, is it `the map` that is `of length 32`, or is it the `byte arrays` that are `of length 32`? I'll revise this suggestion, if I've guessed wrong.
```suggestion
Initialize |components| to an array that is the result of CBOR-decoding the bytes
that follow the three-byte ECDSA-SD disclosure proof header. If the result is not
an array of the following five elements — a byte array of length 64; a byte array
of length 36; an array of byte arrays, each of length 64; a map of integers to
byte arrays, each of length 32; and an array of integers — an error MUST be raised
and SHOULD convey an error type of `PROOF_VERIFICATION_ERROR`.
```

File: index.html [hunk 13]
```diff
@@ -2469,23 +2476,26 @@ <h4>parseDerivedProofValue</h4>
...
-array of integers, throwing an error if not.
+array of integers, if it does not an error MUST be raised and SHOULD convey an
+error type of `PROOF_VERIFICATION_ERROR`.
```
Comment (TallTed) [inline index.html:2498]:
```suggestion
```

File: index.html [hunk 14]
```diff
@@ -1778,7 +1781,9 @@ <h4>selectPaths</h4>
 Set |parentValue| to |value|.
                 </li>
                 <li>
-Set |value| to |parentValue.|path|. If |value| is now undefined, throw an error
+Set |value| to |parentValue.|path|. If |value| is now undefined,
+an error MUST be raised and SHOULD convey an error type of
+`PROOF_GENERATION_ERROR`
```
Comment (TallTed) [inline index.html:1786]:
```suggestion
`PROOF_GENERATION_ERROR`,
```

File: index.html [hunk 15]
```diff
@@ -2934,9 +2945,10 @@ <h4>Verify Derived Proof (ecdsa-sd-2023)</h4>
 custom JSON-LD API options, such as a document loader.
             </li>
             <li>
-If the length of |signatures| does not match the length of |nonMandatory|, throw
-an error indicating that the signature count does not match the non-mandatory
-message count.
+If the length of |signatures| does not match the length of |nonMandatory|,
+an error MUST be raised and SHOULD convey an error type of
+`PROOF_VERIFICATION_ERROR` indicating that the signature count does not match
```
Comment (TallTed) [inline index.html:2950]:
```suggestion
`PROOF_VERIFICATION_ERROR`, indicating that the signature count does not match
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Editorial, punctuation and phrasing, for consistency and clarity

File: index.html (hunk 1 above)
Comment (Wind4Greg) [reply index.html]: I got some weird GitHub errors when trying to batch in all the changes. So had to do it in two batches.

**Review by msporny — APPROVED** [review]

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**